File: 400/457 - Circular Array Loop.c
static int dfs(int *nums, int index, int size, int count) {
    int next;
    
    if (count < size) {
        next = (index + nums[index] + size) % size;
        
        if (next == index || nums[next] * nums[index] <= 0 || dfs(nums, next, size, count + 1) == 0)
            nums[index] = 0;
    }
    
    return nums[index];
}

bool circularArrayLoop(int* nums, int numsSize) {
    int i;
    
    for (i = 0; i < numsSize; ++i) {
        if (nums[i] && dfs(nums, i, numsSize, 0))
            return 1;
    }
    
    return 0;
}

File: csharp/457.circular-array-loop.cs
/*
 * @lc app=leetcode id=457 lang=csharp
 *
 * [457] Circular Array Loop
 */
public class Solution {
    private static int Next(int[] nums, int n, int i) {
        return ((i + nums[i]) % n + n) % n;
    }

    public bool CircularArrayLoop(int[] nums) {
        int n = nums.Length;
        for (int i = 0; i < n; ++i) {
            if (nums[i] == 0) {
                continue;
            }
            int slow = i, fast = Next(nums, n, i), sign = nums[i] > 0 ? 1 : -1;
            while (sign * nums[fast] > 0 && sign * nums[Next(nums, n, fast)] > 0) {
                if (slow == fast) {
                    if (slow != Next(nums, n, slow)) {
                        return true;
                    }
                    break;
                }
                slow = Next(nums, n, slow);
                fast = Next(nums, n, Next(nums, n, fast));
            }
            slow = i;
            while (sign * nums[slow] > 0) {
                fast = Next(nums, n, slow);
                nums[slow] = 0;
                slow = fast;
            }
        }
        return false;
    }
}

